### PR TITLE
Add Niko 552-720X4 Battery switch, 4 button

### PR DESCRIPTION
**Pull request by jovandeginste**
Product page: <https://www.niko.eu/nl-be/artikel/552-72014>

### CODE REVIEW

**Comment by jovandeginste** [comment]
This is my first contribution - please provide guidance.

I see the tests fail, but I'm not sure how to fix this.

**Comment by sjorge** [comment]
The test is failing because the switch expose assumes an access mode of ALL or STATE+GET (I'm on my phone so can't double check)

And you are using fz.command_state_* which is only STATE.

You can fix this by adding a .withAccess(ea.STATE) to each of the exposes.

**Comment by jovandeginste** [comment]
> withAccess(ea.STATE)

When I add this, I get this error:

```
AssertionError [ERR_ASSERTION]: Cannot add access if not defined yet

  27 |
  28 |     withAccess(a) {
> 29 |         assert(this.hasOwnProperty('access'), 'Cannot add access if not defined yet');
     |         ^
  30 |         this.access = a;
  31 |         return this;
  32 |     }
```

**Comment by jovandeginste** [comment]
So, I didn't find any existing example for the `switch`; I tried:

`e.switch().withAccess(ea.STATE).withEndpoint('top_left')`

and

`e.switch().withEndpoint('top_left').withAccess(ea.STATE)`

... but I'm starting to have a feeling that I should use something different than `switch`?

The device sends `commandOn` and `commandOff`, eg:

> Received Zigbee message from '0x...', type 'commandOn', cluster 'genOnOff', data '{}' from endpoint 4 with groupID 0

I also think there should be more information available, but I'm still trying to get to know this stuff...

**Comment by sjorge** [comment]
I think `command_on` and `command_of` might be better suited,  it automatically appends the endpoint name to the action. https://github.com/Koenkk/zigbee-herdsman-converters/blob/92a015f558be1bd5fad1c605c0e52f9ad4e1147b/converters/fromZigbee.js#L1245

So it will publish `{'action': 'on_top_left'}` ...

So something like:
```
    {
        zigbeeModel: ['Battery switch, 4 button'],
        model: '552-720X4',
        vendor: 'NIKO NV',
        description: 'Battery switch, 4 button',
        meta: {multiEndpoint: true},
        fromZigbee: [fz.command_on, fz.command_off],
        toZigbee: [],
        endpoint: (device) => {
            return {top_left: 1, bottom_left: 2, top_right: 3, bottom_right: 4};
        },
        configure: async (device, coordinatorEndpoint, logger) => {
            for (let i = 1; i < 4; i++) {
                const ep = device.getEndpoint(i);
                await reporting.bind(ep, coordinatorEndpoint, ['genOnOff']);
            }
        },
       exposes: [e.action([
            'on_top_left', 'off_top_left',
            'on_bottom_left', 'off_bottom_left',
            'on_top_right', 'off_top_right',
            'on_bottom_right', 'off_bottom_right',
       ])],
    },
```

That seems to be inline with some other battery powered multi button remotes e.g. 
- https://github.com/Koenkk/zigbee-herdsman-converters/blob/92a015f558be1bd5fad1c605c0e52f9ad4e1147b/devices/legrand.js#L113
- https://github.com/Koenkk/zigbee-herdsman-converters/blob/92a015f558be1bd5fad1c605c0e52f9ad4e1147b/devices/legrand.js#L196
- https://github.com/Koenkk/zigbee-herdsman-converters/blob/92a015f558be1bd5fad1c605c0e52f9ad4e1147b/devices/philips.js#L1854

**Comment by jovandeginste** [comment]
I will test this, thanks for your time

**Comment by sjorge** [comment]
Does debug log level still show the commandOn and commandOff messages?

**Comment by jovandeginste** [comment]
I don't understand; I now can't seem to use the device after re-pairing. Is there a way to see anything being received through the Zigbee network?

**Comment by jovandeginste** [comment]
When pairing, the logs show this:

> homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-11 19:02:55: Device '0x...' announced itself
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:info  2023-04-11 19:02:55: MQTT publish: topic 'zigbee2mqtt/bridge/event', payload '{"data":{"friendly_name":"0x...","ieee_address":"0x003c84fffe8be5e9"},"type":"device_announce"}'
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-11 19:02:57: Retrieving state of '0x...' after reconnect
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-11 19:03:22: Received Zigbee message from '0x...', type 'commandQueryNextImageRequest', cluster 'genOta', data '{"fieldControl":0,"fileVersion":553936736,"imageType":43534,"manufacturerCode":4703}' from endpoint 1 with groupID 0
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-11 19:03:22: Device '0x...' requested OTA
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-11 19:03:22: Responded to OTA request of '0x...' with 'NO_IMAGE_AVAILABLE'

When I press buttons, no events are recorded...

**Comment by sjorge** [comment]
Debug should indeed list everything z2m receives from the device, there doesnt seem to be anything in your excerpt above.

Does it show anything when you revert to your original one?

**Comment by jovandeginste** [comment]
Something was wrong with my for-loop; this was indeed something I added at the last moment, yesterday evening...

**Comment by jovandeginste** [comment]
Instead of state switches like 
image
... there is now an "action" sensor.

This seems to work, and you can attach actions to it. Is this how it's supposed to work?

**Comment by jovandeginste** [comment]
The device is wireless (battery powered), so I guess it's passive. I'm trying to make sure my code is actually working, now :-)

**Comment by jovandeginste** [comment]
I can't read the attribute:

> error 2023-04-11 20:13:21Publish 'set' 'read' to '0x003c84fffe8be5e9' failed: 'Error: Read 0x003c84fffe8be5e9/1 genOnOff(["onOff"], {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 43761 - 1 - 9 - 6 - 1 after 10000ms)'

**Comment by jovandeginste** [comment]
The current diff in this PR seems to reliably work; it generates 8 types of actions:
image

(unsure if that answers your question)

**Comment by sjorge** [comment]
> I can't read the attribute:
> 
> > error 2023-04-11 20:13:21Publish 'set' 'read' to '0x003c84fffe8be5e9' failed: 'Error: Read 0x003c84fffe8be5e9/1 genOnOff(["onOff"], {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 43761 - 1 - 9 - 6 - 1 after 10000ms)'

OK so that lines up with it just sending commandOn/Off as it has no actually internal relay it switches.
I think using the action here like on a lot of the other battery powered switches would make the most sense.

**Comment by jovandeginste** [comment]
> > I can't read the attribute:
> > > error 2023-04-11 20:13:21Publish 'set' 'read' to '0x003c84fffe8be5e9' failed: 'Error: Read 0x003c84fffe8be5e9/1 genOnOff(["onOff"], {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 43761 - 1 - 9 - 6 - 1 after 10000ms)'
> 
> OK so that lines up with it just sending commandOn/Off as it has no actually internal relay it switches. I think using the action here like on a lot of the other battery powered switches would make the most sense.

Yes, there is no physical relay - no wires go in or out, only a battery :-)

**Comment by jovandeginste** [comment]
> That also seems inline with what to expect, as it's 4 endpoints (buttons) with generate an on/off command.

I had hopes of a "battery" status...

**Comment by jovandeginste** [comment]
image

**Comment by sjorge** [comment]
Looks like it is indeed on endpoint 1 :) try what I mentioned in the previous comment and see if it works, if not. We might need to try via voltage and figure out the correct battery type which is a lot more annoying to get right.

**Comment by jovandeginste** [comment]
I added your code; every "command" now contains: `"battery":100`

Unsure about the accurateness of that number, since it's a brand new battery (well, since yesterday). It's a CR2450 (if that helps in any way)

**Comment by sjorge** [comment]
> I added your code; every "command" now contains: `"battery":100`
> 
> Unsure about the accurateness of that number, since it's a brand new battery (well, since yesterday). It's a CR2450 (if that helps in any way)

Aha that's good, It's using whatever the device itself is reporting for batteryPercentageRemaining.

**Comment by jovandeginste** [comment]
So, it worked, but now I tried to re-pair again and it's no longer working... Again missing those "configuring" lines. Is there any way to see what's wrong in the logs?

**Comment by sjorge** [comment]
> So, it worked, but now I tried to re-pair again and it's no longer working... Again missing those "configuring" lines. Is there any way to see what's wrong in the logs?

With debug logging enable you should see some indication in the log that it is configuring the bendings for genOnOff and the reporting for genPowerCfg.batteryPercentageRemaining.

Could it be that the device went to sleep before z2m was done configuring it?

**Comment by jovandeginste** [comment]
Nothing in the debug logs, so I suspect some kind of syntax error. It's very consistently working without battery, and not working with battery now :(

**Comment by sjorge** [comment]
> Nothing in the debug logs, so I suspect some kind of syntax error. It's very consistently working without battery, and not working with battery now :(

If there is a syntax error it should show during the startup of z2m I think. You can also do:
```
npm install
npm run lint
npm run test
```

In the repo with your changes to have the lint check and tests run which also run as a github action when you push.

**Comment by jovandeginste** [comment]
The linter didn't complain, but the code didn't work anyway.

**Comment by jovandeginste** [comment]
Current code:

```diff
diff --git a/devices/niko.js b/devices/niko.js
index 1444b3ed..0156cd6c 100644
--- a/devices/niko.js
+++ b/devices/niko.js
@@ -311,4 +311,39 @@ module.exports = [
         },
         exposes: [e.cover_position()],
     },
+    {
+        zigbeeModel: ['Battery switch, 4 button'],
+        model: '552-720X4',
+        vendor: 'NIKO NV',
+        description: 'Battery switch, 4 button',
+        meta: {multiEndpoint: true},
+        fromZigbee: [fz.command_on, fz.command_off],
+        toZigbee: [],
+        endpoint: (device) => {
+            return {top_left: 1, bottom_left: 2, top_right: 3, bottom_right: 4};
+        },
+        configure: async (device, coordinatorEndpoint, logger) => {
+            const ep1 = device.getEndpoint(1);
+            await reporting.bind(ep1, coordinatorEndpoint, ['genOnOff']);
+            await reporting.batteryPercentageRemaining(ep1);
+            const ep2 = device.getEndpoint(2);
+            await reporting.bind(ep2, coordinatorEndpoint, ['genOnOff']);
+            const ep3 = device.getEndpoint(3);
+            await reporting.bind(ep3, coordinatorEndpoint, ['genOnOff']);
+            const ep4 = device.getEndpoint(4);
+            await reporting.bind(ep4, coordinatorEndpoint, ['genOnOff']);
+        },
+        exposes: [
+            e.action([
+                'on_top_left',
+                'off_top_left',
+                'on_bottom_left',
+                'off_bottom_left',
+                'on_top_right',
+                'off_top_right',
+                'on_bottom_right',
+                'off_bottom_right',
+            ]),
+        ],
+    },
 ];
```

**Comment by sjorge** [comment]
Try changing:
```
await reporting.bind(ep1, coordinatorEndpoint, ['genOnOff']);
```

to 

```
await reporting.bind(ep1, coordinatorEndpoint, ['genOnOff', 'genPowerCfg']);
```

**Comment by jovandeginste** [comment]
Today I found out how to actually disconnect a device from the network (using `zigbee2mqtt/bridge/request/device/remove`); using that method, I iterated and got to the current diff. I could re-add successfully the same device several times. It now shows a battery symbol (which is always "100%"... might be an issue still?)

**Comment by jovandeginste** [comment]
I'm still wondering if I could do anything with that `genLevelCtrl` (or long pressing, or ...)

**Comment by sjorge** [comment]
Try also binding it and see if more commandX show up in the debug log.~ sjorgeOn 13 Apr 2023, at 00:52, Jo Vandeginste ***@***.***> wrote:﻿
I'm still wondering if I could do anything with that genLevelCtrl (or long pressing, or ...)

File: devices/niko.js [hunk 1]
```diff
@@ -311,4 +311,63 @@ module.exports = [
...
+            return {top_left: 1, bottom_left: 2, top_right: 3, bottom_right: 4};
+        },
+        configure: async (device, coordinatorEndpoint, logger) => {
+            const ep1 = device.getEndpoint(1);
+            await reporting.bind(ep1, coordinatorEndpoint, [
+                'genIdentify',
+                'genGroups',
+                'genOnOff',
+                'genLevelCtrl',
+                'genOta',
```
Comment (sjorge) [inline devices/niko.js:337]: You dont need to bund genOta, probably not genIdentify either.

**Comment by jovandeginste** [comment]
Well, what do I see now!

> homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-13 01:06:47: Received Zigbee message from '0x003c84fffe8be5e9', type 'commandMoveWithOnOff', cluster 'genLevelCtrl', data '{"movemode":0,"rate":50}' from endpoint 2 with groupID 0
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-13 01:06:48: Received Zigbee message from '0x003c84fffe8be5e9', type 'commandStop', cluster 'genLevelCtrl', data '{}' from endpoint 2 with groupID 0
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-13 01:06:50: Received Zigbee message from '0x003c84fffe8be5e9', type 'commandMove', cluster 'genLevelCtrl', data '{"movemode":1,"rate":50}' from endpoint 2 with groupID 0
homeassistant-zigbee2mqtt-1  | Zigbee2MQTT:debug 2023-04-13 01:06:51: Received Zigbee message from '0x003c84fffe8be5e9', type 'commandStop', cluster 'genLevelCtrl', data '{}' from endpoint 2 with groupID 0

Should I now add every combination of ["move", "stop", "commandMoveWithOnOff"], ["left", "right"] and ["top", "bottom"]?

**Comment by jovandeginste** [comment]
whew, that just worked...! (well, the names were different, but still!)

**Comment by jovandeginste** [comment]
afbeelding

Anything else I can do?

**Comment by sjorge** [comment]
Based on the screenshot from the clusters, that’s probably everything

**Comment by jovandeginste** [comment]
> Based on the screenshot from the clusters, that’s probably everything

Yes, but maybe the code can be made cleaner?

**Comment by sjorge** [comment]
> > Based on the screenshot from the clusters, that’s probably everything
> 
> Yes, but maybe the code can be made cleaner?

I think it looks fine, maybe @koenkk has more comments before he merges

**Comment by Koenkk** [comment]
Thanks!

**Comment by norbertbxl** [comment]
Hello
Will this device be added in the next implementation ?

**Comment by norbertbxl** [comment]
Good morning
I noticed that my two switches are not synchronizing the state

if i turn on the lights from another device i have to double click to turn them off

greetings and have a nice weekend

**Comment by jovandeginste** [comment]
Yes, AFAICT the switch is passive (battery powered), therefore it does not synchronize. I added the on/off events as toggles in my tests...

I'm happy to be corrected!

**Comment by norbertbxl** [comment]
you mean to say you added helpers as switches right?

**Comment by sjorge** [comment]
Behavior is similar to e.g. https://www.zigbee2mqtt.io/devices/929002398602.html

when not binding the device to a group i just treat both on and off as the same in my automations. 

When binding to a group, you’re basically out of luck :( even more ‘fun’ if you have multiple bound.

**Comment by jovandeginste** [comment]
I didn't do anything fancy (yet), I treated both the on and off as "toggle power"

**Comment by norbertbxl** [comment]
Ok thanks, I figured it out with helper

**Pull request merged**